util/codec/codec.go

```diff
@@ -284,105 +266,144 @@ func sizeInt(comparable bool) int {
-			b = encodeUnsignedInt(b, val, comparable)
-		case mysql.TypeJSON:
-			b = append(b, jsonFlag)
-			j := row.GetJSON(i)
-			b = append(b, j.TypeCode)
-			b = append(b, j.Value...)
-		default:
-			return nil, errors.Errorf("unsupport column type for encode %d", allTypes[i].Tp)
 		}
+		var v uint64
+		v, err = t.ToPackedUint()
+		if err != nil {
+			return nil, err
+		}
+		b = EncodeUint(b, v)
+	case mysql.TypeDuration:
+		// duration may have negative value, so we cannot use String to encode directly.
+		b = append(b, durationFlag)
+		b = EncodeInt(b, int64(row.GetDuration(idx, 0).Duration))
+	case mysql.TypeNewDecimal:
+		// If hash is true, we only consider the original value of this decimal and ignore it's precision.
+		dec := row.GetMyDecimal(idx)
+		bin, err := dec.ToHashKey()
+		if err != nil {
+			return nil, err
+		}
+		return [][]byte{
+			{decimalFlag},
+			bin,
+		}, nil
+	case mysql.TypeEnum:
+		b = encodeUnsignedInt(b, uint64(row.GetEnum(idx).ToNumber()), comparable)
+	case mysql.TypeSet:
+		b = encodeUnsignedInt(b, uint64(row.GetSet(idx).ToNumber()), comparable)
+	case mysql.TypeBit:
+		// We don't need to handle errors here since the literal is ensured to be able to store in uint64 in convertToMysqlBit.
+		var val uint64
+		val, err = types.BinaryLiteral(row.GetBytes(idx)).ToInt(sc)
+		terror.Log(errors.Trace(err))
+		b = encodeUnsignedInt(b, val, comparable)
+	case mysql.TypeJSON:
+		return [][]byte{
+			{jsonFlag},
+			row.GetBytes(idx),
+		}, nil
+	default:
+		return nil, errors.Errorf("unsupport column type for encode %d", tp.Tp)
 	}
-	return b, errors.Trace(err)
+	return [][]byte{b}, nil
 }
 
-// HashValues appends the encoded values to byte slice b, returning the appended
-// slice. If two datums are equal, they will generate the same bytes.
-func HashValues(sc *stmtctx.StatementContext, b []byte, v ...types.Datum) ([]byte, error) {
-	return encode(sc, b, v, false, true)
+// HashChunkRow writes the encoded values to w.
+// If two rows are equal, it will generate the same bytes.
+func HashChunkRow(sc *stmtctx.StatementContext, w io.Writer, row chunk.Row, allTypes []*types.FieldType, colIdx []int) error {
+	var b []byte
+	for _, idx := range colIdx {
+		rets, err := encodeHashChunkRowIdx(sc, b, row, allTypes[idx], idx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		for _, ret := range rets {
+			_, err = w.Write(ret)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+	return nil
 }
 
-// HashChunkRow appends the encoded values to byte slice "b", returning the appended slice.
-// If two rows are equal, it will generate the same bytes.
-func HashChunkRow(sc *stmtctx.StatementContext, b []byte, row chunk.Row, allTypes []*types.FieldType, colIdx []int) ([]byte, error) {
-	return encodeHashChunkRow(sc, b, row, allTypes, colIdx)
+// EqualChunkRow returns a boolean reporting whether row1 and row2
+// with their types and column index are logically equal.
+func EqualChunkRow(sc *stmtctx.StatementContext,
```

Comment:
You can use `CompareFunc()` to do this, and it does not encode the compare fields. `CompareFunc` is located at `util/chunk/compare.go`

Reply:
`CompareFunc` is good to compare two rows with the same field type in one table, but not good for comparing `uint64` and `int64` or comparing `float` and `double` in different table.

Reply:
How about using `Datum.CompareDatum`?
It should be faster than `EqualChunkRow` since it doesn't encode the data.

Reply:
It's better to do a benchmark to chose the best one.

Reply:
`EqualChunkRow` does not encode the data either now.
I uses `chunk.GetCompareFunc` to benchmark with `EqualChunkRow`, the results are similar.

```
name                                                            old time/op    new time/op    delta
HashJoinExec/(rows:100000,_concurency:4,_joinKeyIdx:_[0_1])-12     730ms ± 3%     732ms ± 2%   ~     (p=0.690 n=5+5)
HashJoinExec/(rows:100000,_concurency:4,_joinKeyIdx:_[0])-12       165ms ± 4%     161ms ± 2%   ~     (p=0.151 n=5+5)

name                                                            old alloc/op   new alloc/op   delta
HashJoinExec/(rows:100000,_concurency:4,_joinKeyIdx:_[0_1])-12     305MB ± 0%     305MB ± 0%   ~     (p=0.151 n=5+5)
HashJoinExec/(rows:100000,_concurency:4,_joinKeyIdx:_[0])-12       305MB ± 0%     305MB ± 0%   ~     (p=0.222 n=5+5)

name                                                            old allocs/op  new allocs/op  delta
HashJoinExec/(rows:100000,_concurency:4,_joinKeyIdx:_[0_1])-12      306k ± 0%      306k ± 0%   ~     (p=0.333 n=5+5)
HashJoinExec/(rows:100000,_concurency:4,_joinKeyIdx:_[0])-12        306k ± 0%      306k ± 0%   ~     (p=0.421 n=5+5)
```